packages/core/src/components/panel-stack/panelStack.tsx

```diff
@@ -63,46 +63,36 @@ export class PanelStack extends React.PureComponent<IPanelStackProps, IPanelStac
 
     private renderCurrentPanel() {
         const { stack } = this.state;
+        if (stack.length === 0) {
+            return null;
+        }
         const [activePanel, previousPanel] = stack;
         return (
             <CSSTransition classNames={Classes.PANEL_STACK} key={stack.length} timeout={400}>
                 <PanelView
-                    key={stack.length}
-                    previousPanel={previousPanel}
+                    onClose={this.handlePanelClose}
+                    onOpen={this.handlePanelOpen}
                     panel={activePanel}
-                    panelProps={this.getPanelProps()}
+                    previousPanel={previousPanel}
                 />
             </CSSTransition>
         );
     }
 
-    private getPanelProps(): IPanelProps {
-        const stackSize = this.state.stack.length;
-        return {
-            closePanel: () => this.closePanel(stackSize),
-            openPanel: this.openPanel,
-        };
-    }
-
-    private closePanel = (stackSize: number) => {
-        if (this.state.stack.length !== stackSize || this.state.stack.length <= 1) {
+    private handlePanelClose = (panel: IPanel) => {
+        const { stack } = this.state;
+        // only remove this panel if it is at the top and not the only one.
+        if (stack[0] !== panel || stack.length === 0) {
             return;
         }
-        this.setState(state => {
-            safeInvoke(this.props.onClose, state.stack[0]);
-            return {
-                direction: "pop",
-                stack: state.stack.slice(1),
-            };
-        });
+        safeInvoke(this.props.onClose, panel);
+        this.setState(state => ({
+            direction: "pop",
+            stack: state.stack.filter(p => p !== panel),
```

Comment:
Can we not `slice(1)` here again? If the panel is not at the top of the stack, it won't get here either way?

Reply:
it's not necessarily at the top inside this callback as it's async. the filter is ultimately the most robust check.